### PR TITLE
Integrate deep cache

### PR DESCRIPTION
**Pull request by clackhan**
## ComfyUI Node and Example
Node name: ModuleDeepCacheSpeedup

Example workflow: https://github.com/Oneflow-Inc/onediff/blob/integrate_deep_cache/onediff_comfy_nodes/workflows/deep-cache.png

## Depending on the new oneflow community edition

cuda 11.8: `python3 -m pip install --pre oneflow -f https://oneflow-pro.oss-cn-beijing.aliyuncs.com/branch/community/cu118`

cuda12.1: `python3 -m pip install --pre oneflow -f https://oneflow-pro.oss-cn-beijing.aliyuncs.com/branch/community/cu121`

cuda12.2: `python3 -m pip install --pre oneflow -f https://oneflow-pro.oss-cn-beijing.aliyuncs.com/branch/community/cu122`

### a100
- 1024 * 1024

|SDXL(Pytorch)| DeepCache | SDXL(with OneDiff) |DeepCache with OneDiff|
| :-: | :-: | :-: | :-: |
|7.79it/s(1x)| 16.92it/s(2.17x) | 10.11it/s(1.30x)  |  23.58it/s(3.03x) |

-  512 * 512

|SDXL(Pytorch)| DeepCache | SDXL(with OneDiff) |DeepCache with OneDiff|
| :-: | :-: | :-: | :-: |
|22.47it/s(1x)| 48.73it/(2.17x) | 28.67it/s(1.26x)  |  54.79it/s(2.44x) |

### 3090 
- 1024 * 1024

|SDXL(Pytorch)| DeepCache | SDXL(with OneDiff) |DeepCache with OneDiff|
| :-: | :-: | :-: | :-: |
|4.01it/s(1x)|  8.79it/s(2.19x) | 6.41it/s(1.6x)  |  14.11it/s(3.52x) |

-  512 * 512

|SDXL(Pytorch)| DeepCache | SDXL(with OneDiff) |DeepCache with OneDiff|
| :-: | :-: | :-: | :-: |
|13.88it/(1x)|  29.81it/s(2.15x) | 21.49it/s(1.55x)  |  43.56it/s(3.14x) |

### CODE REVIEW

File: onediff_comfy_nodes/__init__.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
     ControlNetGraphLoader,
     ControlNetGraphSaver,
     SVDSpeedup,
+    ModelDeepCache,
+    OfModuleDeepCache,
```
Comment (strint) [inline onediff_comfy_nodes/__init__.py:15]: ModuleDeepCacheSpeedup

File: onediff_comfy_nodes/_nodes.py [hunk 2]
```diff
@@ -402,3 +404,342 @@ def quantize_model(self, model, output_dir, conv, linear):
...
+            :param x: an [N x C x ...] Tensor of inputs.
+            :param timesteps: a 1-D batch of timesteps.
+            :param context: conditioning plugged in via crossattn
+            :param y: an [N] Tensor of labels, if class-conditional.
+            :return: an [N x C x ...] Tensor of outputs.
+            """
+            unet = new_model.model.diffusion_model
+
+
+            # unet次回実行はtimestepが上がってると仮定・・Refiner等でエラーが起きるかも
```
Comment (strint) [inline onediff_comfy_nodes/_nodes.py:656]: 这种类型的注释可以去掉或者翻译下

File: onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                patch = transformer_patches["input_block_patch_after_skip"]
+                for p in patch:
+                    h = p(h, transformer_options)
+            if id == self.cache_depth:
+                break
+
+        # transformer_options["block"] = ("middle", 0)
+        # h = forward_timestep_embed(
+        #     self.unet_module.middle_block, h, emb, context, transformer_options, time_context=time_context, num_video_frames=num_video_frames, image_only_indicator=image_only_indicator)
+        # h = apply_control(h, control, 'middle')
```
Comment (strint) [inline onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py:152]: 这些没用？

File: onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                for p in patch:
+                    h, hsp = p(h, hsp, transformer_options)
+
+            h = torch.cat([h, hsp], dim=1)
+            del hsp
+            if len(hs) > 0:
+                output_shape = hs[-1].shape
+            else:
+                output_shape = None
+            h = forward_timestep_embed(module, h, emb, context, transformer_options, output_shape, time_context=time_context, num_video_frames=num_video_frames, image_only_indicator=image_only_indicator)
```
Comment (strint) [inline onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py:177]: 这里看起来应该 format 一下

File: onediff_comfy_nodes/_nodes.py [hunk 5]
```diff
@@ -402,3 +404,342 @@ def quantize_model(self, model, output_dir, conv, linear):
...
+
+start_step, end_step: apply this method when the timestep is between start_step and end_step
+cache_interval: interval of caching (1 means no caching)
+cache_depth: depth of caching
+'''
+
+import torch
+from comfy.ldm.modules.diffusionmodules.openaimodel import forward_timestep_embed, timestep_embedding, th, apply_control
+
+class ModelDeepCache:
```
Comment (strint) [inline onediff_comfy_nodes/_nodes.py:562]: 这个注释下主干代码来源于：https://gist.github.com/laksjdjf/435c512bc19636e9c9af4ee7bea9eb86

File: onediff_comfy_nodes/_nodes.py [hunk 6]
```diff
@@ -402,3 +404,342 @@ def quantize_model(self, model, output_dir, conv, linear):
...
+
+                hs.append(h)
+                if "input_block_patch_after_skip" in transformer_patches:
+                    patch = transformer_patches["input_block_patch_after_skip"]
+                    for p in patch:
+                        h = p(h, transformer_options)
+                
+                if id == cache_depth and apply: 
+                    if not current_step % cache_interval == 0:
+                        break # cache位置以降はスキップ
```
Comment (strint) [inline onediff_comfy_nodes/_nodes.py:700]: 这种日文注释需转成英文

File: onediff_comfy_nodes/__init__.py (hunk 1 above)
Comment (clackhan) [reply onediff_comfy_nodes/__init__.py:15]: > ModuleDeepCacheSpeedup

已修改

File: onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py (hunk 3 above)
Comment (clackhan) [reply onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py:152]: > 这些没用？

没用，已删除

File: onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py (hunk 4 above)
Comment (clackhan) [reply onediff_comfy_nodes/infer_compiler_registry/register_comfy/deep_cache_unet.py:177]: > 这里看起来应该 format 一下

已经format

File: onediff_comfy_nodes/_nodes.py (hunk 5 above)
Comment (clackhan) [reply onediff_comfy_nodes/_nodes.py]: > 这个注释下主干代码来源于：https://gist.github.com/laksjdjf/435c512bc19636e9c9af4ee7bea9eb86 致谢一下

pytorch 版的 deepcache 仅作为对比使用，对齐之后没有必要保留，已删除。 并在oneflow 版 deepcache node中注明了参考来源

File: onediff_comfy_nodes/_nodes.py (hunk 6 above)
Comment (clackhan) [reply onediff_comfy_nodes/_nodes.py]: > 这种日文注释需转成英文

已转换

File: onediff_comfy_nodes/__init__.py [hunk 7]
```diff
@@ -43,6 +45,8 @@
     "ControlNetGraphLoader": "ControlNet Graph Loader",
     "ControlNetGraphSaver": "ControlNet Graph Saver",
     "SVDSpeedup": "SVD Speedup",
+    "SVDSpeedup": "SVD Speedup",
```
Comment (strint) [inline onediff_comfy_nodes/__init__.py:48]: 这里重复了

**Review by strint — APPROVED** [review]

File: onediff_comfy_nodes/__init__.py (hunk 7 above)
Comment (clackhan) [reply onediff_comfy_nodes/__init__.py]: > 这里重复了

已修改

**Pull request merged**